File: samples/CameraAccess/CameraAccess/Display/DisplayViewModel.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

//
// DisplayViewModel.swift
//
// Drives the MWDATDisplay capability (new in SDK 0.7.0). It selects a
// display-capable device, starts a DeviceSession, attaches a Display, and sends
// view content to Meta Ray-Ban Display glasses. Display requires the Device
// Access Toolkit App Model — the app opts in via MWDAT.DAMEnabled in Info.plist.
//

import Combine
import MWDATCore
import MWDATDisplay

@MainActor
final class DisplayViewModel: ObservableObject {
  @Published var isConnecting: Bool = false
  @Published var isConnected: Bool = false
  @Published var statusMessage: String = ""
  @Published var showError: Bool = false
  @Published var errorMessage: String = ""

  private let wearables: WearablesInterface
  private let deviceSelector: AutoDeviceSelector
  private var deviceSession: DeviceSession?
  private var display: Display?
  private var displayStateToken: AnyListenerToken?
  private var sessionErrorTask: Task<Void, Never>?
  // Content queued while the display session is still coming up; sent on `.started`.
  private var pendingView: FlexBox?

  init(wearables: WearablesInterface) {
    self.wearables = wearables
    // Only auto-select devices that can render a display (e.g. Ray-Ban Display).
    self.deviceSelector = AutoDeviceSelector(wearables: wearables, filter: { $0.supportsDisplay() })
  }

  deinit {
    displayStateToken = nil
    sessionErrorTask?.cancel()
    deviceSession?.stop()
  }

  // MARK: - Public API

  /// Sends a "Hello World" card to the glasses display, attaching a display
  /// session first if one isn't already connected.
  func sendHelloWorld() async {
    await send(HelloWorldDisplay.helloWorld())
  }

  /// Stops the Display capability and the underlying device session.
  func disconnect() async {
    displayStateToken = nil
    sessionErrorTask?.cancel()
    sessionErrorTask = nil
    await display?.stop()
    display = nil
    deviceSession?.stop()
    deviceSession = nil
    pendingView = nil
    isConnected = false
    isConnecting = false
    statusMessage = ""
  }

  func dismissError() {
    showError = false
    errorMessage = ""
  }

  // MARK: - Private

  private func send(_ view: FlexBox) async {
    if let display, isConnected {
      await doSend(view, on: display)
      return
    }
    // Not connected yet: queue the content and bring the display session up.
    pendingView = view
    if display == nil {
      await attachToDisplay()
    }
  }

  private func doSend(_ view: FlexBox, on display: Display) async {
    do {
      // Each send replaces the previous root view shown on the glasses.
      try await display.send(view)
      statusMessage = "Sent \u{201C}Hello World\u{201D} to the glasses display"
    } catch {
      handleError((error as? DisplayError)?.description ?? error.localizedDescription)
    }
  }

  private func attachToDisplay() async {
    guard display == nil else { return }
    isConnecting = true
    statusMessage = "Connecting to display\u{2026}"

    do {
      let session = try wearables.createSession(deviceSelector: deviceSelector)
      deviceSession = session

      let stateStream = session.stateStream()
      let errorStream = session.errorStream()

      // Surface asynchronous session failures.
      sessionErrorTask = Task { [weak self] in
        for await error in errorStream {
          guard let self else { return }
          self.handleError(error.localizedDescription)
        }
      }

      try session.start()

      // The session may reach `.started` before the for-await begins iterating.
      if session.state == .started {
        await setupDisplay(on: session)
        return
      }
      for await state in stateStream {
        if state == .started {
          await setupDisplay(on: session)
          return
        }
        if state == .stopped {
          handleError("The device session stopped before the display could start.")
          return
        }
      }
    } catch DeviceSessionError.datAppOnTheGlassesUpdateRequired {
      handleError(DeviceSessionError.datAppOnTheGlassesUpdateRequired.localizedDescription)
    } catch DeviceSessionError.noEligibleDevice {
      handleError("No Meta Ray-Ban Display glasses found. Connect display-capable glasses and try again.")
    } catch {
      handleError("Failed to start display session: \(error.localizedDescription)")
    }
  }

  private func setupDisplay(on session: DeviceSession) async {
    guard display == nil else { return }
    do {
      let capability = try session.addDisplay()
      display = capability
      displayStateToken = capability.statePublisher.listen { [weak self] state in
        Task { @MainActor [weak self] in
          await self?.handleDisplayState(state, on: capability)
        }
      }
      await capability.start()
    } catch {
      handleError("Failed to attach display: \(error.localizedDescription)")
    }
  }

  private func handleDisplayState(_ state: DisplayState, on display: Display) async {
    switch state {
    case .starting:
      break
    case .started:
      isConnecting = false
      isConnected = true
      statusMessage = "Display connected"
      // Flush any content the user requested before the display was ready.
      if let view = pendingView {
        pendingView = nil
        await doSend(view, on: display)
      }
    case .stopping:
      isConnected = false
    case .stopped:
      isConnected = false
      displayStateToken = nil
      self.display = nil
      sessionErrorTask?.cancel()
      sessionErrorTask = nil
      deviceSession?.stop()
      deviceSession = nil
      isConnecting = false
    }
  }

  private func handleError(_ message: String) {
    errorMessage = message
    showError = true
    statusMessage = ""
    isConnecting = false
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/display/DisplayViewModel.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.meta.wearable.dat.externalsampleapps.cameraaccess.display

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.AutoDeviceSelector
import com.meta.wearable.dat.core.session.DeviceSession
import com.meta.wearable.dat.core.session.DeviceSessionState
import com.meta.wearable.dat.core.types.DeviceSessionError
import com.meta.wearable.dat.display.Display
import com.meta.wearable.dat.display.addDisplay
import com.meta.wearable.dat.display.types.DisplayConfiguration
import com.meta.wearable.dat.display.types.DisplayState
import com.meta.wearable.dat.display.views.ContentScope
import com.meta.wearable.dat.externalsampleapps.cameraaccess.display.DisplayWidgets.board
import com.meta.wearable.dat.externalsampleapps.cameraaccess.display.DisplayWidgets.helloWorld
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class DisplayUiState(
    val isConnecting: Boolean = false,
    val isConnected: Boolean = false,
    val statusMessage: String = "",
    val showError: Boolean = false,
    val errorMessage: String = "",
)

class DisplayViewModel(application: Application) : AndroidViewModel(application) {
    companion object {
        private const val TAG = "DisplayViewModel"
    }

    private val _uiState = MutableStateFlow(DisplayUiState())
    val uiState: StateFlow<DisplayUiState> = _uiState.asStateFlow()

    private val deviceSelector = AutoDeviceSelector(filter = { it.isDisplayCapable() })
    private var deviceSession: DeviceSession? = null
    private var display: Display? = null
    private var sessionStateJob: Job? = null
    private var sessionErrorJob: Job? = null
    private var displayStateJob: Job? = null
    private var pendingContent: (ContentScope.() -> Unit)? = null

    fun sendHelloWorld() {
        send { helloWorld() }
    }

    fun sendBoard(specs: List<WidgetSpec>) {
        if (specs.isEmpty()) return
        send { board(specs) }
    }

    fun disconnect() {
        displayStateJob?.cancel()
        sessionStateJob?.cancel()
        sessionErrorJob?.cancel()
        display?.stop()
        display = null
        deviceSession?.stop()
        deviceSession = null
        pendingContent = null
        _uiState.value = DisplayUiState()
    }

    fun dismissError() {
        _uiState.update { it.copy(showError = false, errorMessage = "") }
    }

    private fun send(content: ContentScope.() -> Unit) {
        if (display != null && _uiState.value.isConnected) {
            viewModelScope.launch { doSend(content) }
            return
        }
        pendingContent = content
        if (display == null) {
            attachToDisplay()
        }
    }

    private suspend fun doSend(content: ContentScope.() -> Unit) {
        val activeDisplay = display ?: return
        activeDisplay.sendContent(content).fold(
            onSuccess = {
                _uiState.update {
                    it.copy(statusMessage = "Sent content to the glasses display")
                }
            },
            onFailure = { error, _ ->
                handleError(error.description)
            },
        )
    }

    private fun attachToDisplay() {
        if (display != null) return
        _uiState.update {
            it.copy(isConnecting = true, statusMessage = "Connecting to display…")
        }

        Wearables.createSession(deviceSelector)
            .onSuccess { session ->
                deviceSession = session
                sessionErrorJob = viewModelScope.launch {
                    session.errors.collect { error -> handleError(error.description) }
                }
                session.start()
                sessionStateJob = viewModelScope.launch {
                    session.state.collect { state ->
                        when (state) {
                            DeviceSessionState.STARTED -> setupDisplay(session)
                            DeviceSessionState.STOPPED ->
                                handleError(
                                    "The device session stopped before the display could start."
                                )
                            else -> Unit
                        }
                    }
                }
            }
            .onFailure { error, _ ->
                val message =
                    when (error) {
                        DeviceSessionError.DAT_APP_ON_THE_GLASSES_UPDATE_REQUIRED ->
                            error.description
                        DeviceSessionError.NO_ELIGIBLE_DEVICE ->
                            "No Meta Ray-Ban Display glasses found. Connect display-capable glasses and try again."
                        else -> "Failed to start display session: ${error.description}"
                    }
                handleError(message)
            }
    }

    private fun setupDisplay(session: DeviceSession) {
        if (display != null) return
        viewModelScope.launch {
            session.addDisplay(DisplayConfiguration())
                .onSuccess { capability ->
                    display = capability
                    displayStateJob = viewModelScope.launch {
                        capability.state.collect { state -> handleDisplayState(state) }
                    }
                }
                .onFailure { error, _ ->
                    handleError("Failed to attach display: ${error.description}")
                }
        }
    }

    private fun handleDisplayState(state: DisplayState) {
        when (state) {
            DisplayState.STARTING -> Unit
            DisplayState.STARTED -> {
                _uiState.update {
                    it.copy(
                        isConnecting = false,
                        isConnected = true,
                        statusMessage = "Display connected",
                    )
                }
                pendingContent?.let { content ->
                    pendingContent = null
                    viewModelScope.launch { doSend(content) }
                }
            }
            DisplayState.STOPPING -> _uiState.update { it.copy(isConnected = false) }
            DisplayState.STOPPED, DisplayState.CLOSED -> {
                displayStateJob?.cancel()
                sessionStateJob?.cancel()
                sessionErrorJob?.cancel()
                display = null
                deviceSession?.stop()
                deviceSession = null
                pendingContent = null
                _uiState.update {
                    it.copy(isConnecting = false, isConnected = false, statusMessage = "")
                }
            }
        }
    }

    private fun handleError(message: String) {
        Log.e(TAG, message)
        _uiState.update {
            it.copy(
                errorMessage = message,
                showError = true,
                statusMessage = "",
                isConnecting = false,
            )
        }
    }

    override fun onCleared() {
        disconnect()
        super.onCleared()
    }
}
